controllers/clusterca_to_secret.go

```diff
@@ -0,0 +1,190 @@
+
+	ticker := time.NewTicker(20 * time.Second)
+	quit := make(chan struct{})
+	for {
+		select {
+		case <-ticker.C:
+
+			isPresent, err := checkIfSecretPresent(context.Background(), client, log)
+
+			if isPresent {
+				// check the hash of the secret data with cert.
+				sec, err := getSecret(context.Background(), client, log)
+				if err != nil {
+					log.Error(err, "Failed to get secret")
+				}
+
+				clusterCert, err := getCert(log)
+				if err != nil {
+					log.Error(err, "Failed to get cluster CA")
+				}
+
+				clusterCAPemCertHash, err := getPemCertificateHash(*clusterCert, log)
+				if err != nil {
+					log.Error(err, "Failed to get cluster CA Hash")
+				}
+
+				if clusterCAPemCertHash == getCertificateHashFromSecret(*sec) {
+					log.Info("Cert and Secret Hash Match, do nothing...")
+				} else {
+					log.Info("Cert and Secret Hash do not match, update existing")
+					updateSecret(context.Background(), client, log)
+				}
+
+			} else {
+
+				if err != nil {
+					log.Info("Secret Not Found, create new secret")
+				}
+
+				err := createSecret(context.Background(), client, log)
+				if err != nil {
+					log.Error(err, "Error creating secret")
+				}
+
+			}
+
+		case <-quit:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+func getSecret(ctx context.Context, client client.Client, log logr.Logger) (*v1.Secret, error) {
+	var secret v1.Secret
+
+	err := client.Get(ctx, types.NamespacedName{Namespace: Namespace, Name: SecretName}, &secret)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secret: %s", err.Error())
+	}
+
+	return &secret, nil
+}
+
+func checkIfSecretPresent(ctx context.Context, client client.Client, log logr.Logger) (bool, error) {
+	var secret v1.Secret
+
+	err := client.Get(ctx, types.NamespacedName{Namespace: Namespace, Name: SecretName}, &secret)
+
+	if err != nil {
+		log.Error(err, "Error while checking for secret")
+	} else {
+		return true, err
+	}
+	return false, err
+}
+
+func updateSecret(ctx context.Context, client client.Client, log logr.Logger) {
+	log.Info("Update Secret Function")
+
+	var secret v1.Secret
+	sec := client.Get(ctx, types.NamespacedName{Namespace: Namespace, Name: SecretName}, &secret)
+	if sec != nil {
+		log.Error(sec, "Error while checking for secret")
+	} else {
+		err := client.Delete(ctx, &secret)
+
+		if err != nil {
+			log.Error(err, "Unable to delete secret")
+		}
+
+		createSecret := createSecret(ctx, client, log)
+		if createSecret != nil {
+			log.Error(createSecret, "Error creating secret")
+		}
+
+	}
+}
+
+func createSecret(ctx context.Context, client client.Client, log logr.Logger) error {
+
+	fmt.Println("Creating New Secret")
+	var secret v1.Secret
+
+	if err := client.Get(ctx, types.NamespacedName{Namespace: Namespace, Name: SecretName}, &secret); err != nil {
+		if errors.IsNotFound(err) {
+			secret = v1.Secret{
+				ObjectMeta: v12.ObjectMeta{
+					Namespace:   Namespace,
+					Name:        SecretName,
+					Annotations: map[string]string{},
```

Comment:
For annotations, we'll need 'trustbuilder.directv.com/trustedcertificate: "true"' to ensure the secret gets picked up by the controller for adding to truststores. Also, we'll need to update the certificatepackage controller so that it loads the cluster-ca secret regardless of the label-selector the certificatePackage object has. This will likely mean we'll want a cluster-ca specific label on this secret. Maybe "trustbuilder-global=cluster-ca". We will need to update that controller to handle this. I'm looking over that section of code to find suggestion for best way to approach that.

Reply:
Look for that cluster ca secret instead of pulling from the file for Cluster CA.